### DISCUSSION TITLE
Citations with Bibtex

### DISCUSSION THREAD

**Discussion by lambtho12** [nvim-telekasten/telekasten.nvim#82] [Ideas]
I just found out about [telescope-bibtex](https://github.com/nvim-telescope/telescope-bibtex.nvim).

I have not tried it (yet), but if it works as expected, it should be fairly easy to wrap Telekasten around it and use that as a way to easily paste citations in notes.  And if wrapping telekasten around it is too complicated, the extension is under MIT license, so we could incorporate it directly within Telekasten.

What do you think?

**Comment by jglueckstein** [comment]
Wow.  I would love this feature.

**Comment by lambtho12** [comment]
Just tried [the plugin](https://github.com/nvim-telescope/telescope-bibtex.nvim). It works really well, even with a large number of entries (+350). It evens allows for customization of the cite format!

Although, so far I feel it is a bit limited... In the workflow I have in mind, I would use this plugin to add the citekey in the text (taking advantage of the custom format to generate a wiki-style ref)
```markdown
## My note
Lorem ipsum dolor sit amet, consectetur adipiscing elit.[@citekey]. 
```

but then with some magic command (_I have absolutely no idea how_), it would generate a human-readble citation (APA, Chicago, whatever) at the bottom of the document.

```markdown
## My note
Lorem ipsum dolor sit amet, consectetur adipiscing elit.[@citekey]. 

## References (auto-generated)
[@citekey]: Name F., 2022, "My awesome source", Journal of Neovim, Paris, France
```

that would be so cool. Although the plugin does not currently support these types of citations.

**Reply by sbock17** [reply]
As you might know, when rendering the document via pandoc, it is totally possible to have the references renderd, at least if you pass the location of your bibliography in the pandoc command. [Here is an example which renders many different plaintext files to pdf format](https://github.com/LukeSmithxyz/voidrice/blob/d76f894f846dea00b3c568ee8f4a5105686fd559/.local/bin/compiler), which can be tweaked a lot, where the compiler script is defined in the vimrc as
```vim
" Compile document, be it groff/LaTeX/markdown/etc.
	map <leader>c :w! \| !compiler <c-r>%<CR>
```

But this would just come down to add references in a seperate compiled pdf, not directly in plaintext. For the idea of using  vim as IDE, this would be quite nicer !!

It should be very doable however, e.g. using citation.vim. I'll keep my eyes open for this and post if I find some solution. Nevertheless, thanks for recommending telescope-bibtex, just few weeks ago I tried through several options and stood still by the one that worked, e.g. I'm currently using [ZoteroCite](https://retorque.re/zotero-better-bibtex/citing/cayw/),  zotero just needs to be opened somewhere, when I type <leader>Z, then I can just type some key and it gets completed via selection menu, somehow toggled from zotero. That's a very primitive way, which just works. 
There is also [citation.vim ](https://github.com/rafaqz/citation.vim), which for me also works but a bit buggy. Since you can read a bib file with it, and it can output various info for it, there would lie  the solution probably. 
 [zotcite](https://github.com/jalvesaq/zotcite) could also provide some soltuion maybe, but it slowed down vim too much to be usable (at least for me, and maybe my configs are just messed up aswell). It seems like I've overlooked telescope-bibtex, will try it out the next days since I have to write a bunch of texts with references the coming weeks :)

**Reply by lambtho12** [reply]
> But this would just come down to add references in a seperate compiled pdf, not directly in plaintext. For the idea of using vim as IDE, this would be quite nicer !!

Exactly, I do not care at all about generating pdfs. In my workflow, I want to stay within vim, with everything in markdown. If I wanted pdfs with good support for bibligraphies, I would directly type in LaTex...

---
I feel that it should be _fairly easy_ to add a formatter to telescope-bibtex. So far it is capable of outputting either the citekey or the complete bibtex entry. The only missing point would be to pass the bibtex entry to a formatter and allow that to be pasted. However, I think this function should be part of telescope-bibtex, and not performed by telekasten.

After that, it is _just_ a question of looping over all `[@xxxx]` and pasting it all (this should be handled by telekasten).

It will take a bit of time to develop, but it does not seems to be an overly difficult task at first glance.

**Reply by lambtho12** [reply]
> I feel that it should be fairly easy to add a formatter to telescope-bibtex

So.... I have been busy :smile: 
[Telescope-bibtex](https://github.com/nvim-telescope/telescope-bibtex.nvim) pickers can now output formatted citations (see https://github.com/nvim-telescope/telescope-bibtex.nvim/pull/33). You can even add a template to customize the output.

I think the next step would be to add a picker that take in input the word under cursor (like the "follow_link" here) in telescope-bibtex.

**Reply by renerocksai** [reply]
This is awesome work!!!!!!! I remember implementing `# References (auto-generated)` in Sublimeless_ZK quite some time ago. I used pandoc to generate them - and some regex magic to parse the bib file. If it helps: [here](https://github.com/renerocksai/sublimeless_zk/blob/master/src/autobib.py) is the src.

The Sublimeless_ZK workflow in action (could be replaced by telescope-bibtex.nvim and some telekasten "generate bib refs" command)

![](https://raw.githubusercontent.com/renerocksai/sublimeless_zk/6738375c0e371f0c2fde0aa9e539242cfd2b4777/imgs/auto-bib.gif)

**Reply by lambtho12** [reply]
@renerocksai , yes I know. I skimmed you Sublimeless_ZK project when doing that to see how you did. However, I think it is preferable to limit the external dependencies as much as possible. Here, the telekasten_bibtex plugin only is just a tiny telescope add-on and is contained within nvim. I try to avoid pandoc as much as possible, because it is "heavy" and requires the installation of a ton of haskell stuff on Arch Linux.

I did not developed this feature more than that because you were not there, so I did not want to lose my time... Now that I am a contributor (thanks btw), I may finish what I have in mind. I am fairly busy as well at the moment though, so it might take a while.
